### ISSUE TITLE
Server for the new Frontend

### ISSUE THREAD

**Issue by wellemut** [kartevonmorgen/kartevonmorgen.ts#2]
Markus Could you grant Access for @navid_kalaei to our server via SSH with rights to install dependencies, (Yarn) be able to execute applications (pn2), Continous integration and continous deplyment?
Please for two domains https://test.kartevonmorgen.org/ (internal testing)  https://beta.kartevonmorgen.org/ (testing with clients)

**Comment by wellemut** [comment]
Navid Kalaei, [10.03.21 22:34]
oh no, we don't need to setup a new server completely from zero, it's just a node app that can be run beside other applications

Markus Kohlhase, [10.03.21 22:41]
@navid_kalaei can you create a Nix Package for your service? Then it should be easy to integrate

Markus Kohlhase, [10.03.21 22:41]
https://github.com/svanderburg/node2nix

Markus Kohlhase, [10.03.21 22:42]
here you probably can find some upstream examples: https://github.com/NixOS/nixpkgs

Navid Kalaei, [10.03.21 22:49]
How it's differentiated from containers?

Navid Kalaei, [10.03.21 22:52]
And how I can test that??? Couldn't we setup a GitHub Action to make life easier?

**Comment by wellemut** [comment]
- Nix-Package schreiben
- Eigener neuer Server NodeJS
- NodeJS runtime in der cloud
- GitHub Action hat damit nichts zu tun...

**Comment by wellemut** [comment]
@navid-kalaei 
If you know NixOS you could build a Package and send it to @flosse . If you feel comftable with, you can also manage the whole NixOS Server by yourself.

But If you prefer a different system, let me know what system we can buy for you. Only limitation: All Servers must be in Germany. So what would be the german alternative to https://vercel.com/new?utm_source=create-next-app&utm_medium=default-template&filter=next.js&utm_campaign=create-next-app-readme?

**Comment by wellemut** [comment]
I will by this server: https://www.netcup.eu/bestellen/produkt.php?produkt=2005
@flosse please provide me with our netcup credentials

@dmitrap do you have your own server in the meantime? 
It would be no problem for us, if you use this server too

**Comment by dmitrap** [comment]
that's pretty cool, maybe we'll migrate after 3 month that i still paid for
our own server! :)

пт, 3 сент. 2021 г. в 21:17, Helmut Wolman ***@***.***>:

**Comment by navid-kalaei** [comment]
thank you very much @wellemut for your research. there are some points that needs to get cleared.

1. NixOS is not related to the front-end. Any OS including Ubuntu which is more available from the hosting websites is capable of running Node.js and as the consequence Next.js
1. I suggest to purchase a server with a more popular OS rather than NixOS since it's a very uncommon operating system and it's benefits are not sensible. not only me but also future developers will face problems with their deployment. as the result the deployment procedure will be deligated to SlowTec automatically which will comes with inter-organization calls and delays 
1. SSH is compulsory since with our requirements exporting static HTMLs is not possible. (references are provided below)
1. As the time I am writing this comment, no sensitive data will be stored on the front-end server
1. The term defending the security has many meanings. most of the standard data centers protect the servers and backup their data regularly, could you please explain more about the security flaws you are concerned with so we can have a more precise discussion and find solutions if necessary?
1. I can setup the server for the front-end on a normal operating system if you want

**Comment by wellemut** [comment]
@alex0107 Offers to do our hosting of backend and frontend on a Server he has ordered from [Hetzner.](https://www.hetzner.com/cloud)

@navid-kalaei Could you please make sure, that all Server-Dependencies are correctly and fully described here: https://github.com/kartevonmorgen/kartevonmorgen.ts#test-platform (Including OS, Node.js ...)

@flosse Can you please describe all Server-Requirements here https://github.com/kartevonmorgen/openfairdb/#build

Alex will provide Debian on his server and will give you test access. Do you think this could work?

**Comment by alex0107** [comment]
@navid-kalaei is it possible to start the new frontend with node.js instead of pm2 or can we create and use a Docker-Image?

**Comment by navid-kalaei** [comment]
> @alex0107 Offers to do our hosting of backend and frontend on a Server he has ordered from [Hetzner.](https://www.hetzner.com/cloud)
> 
> @navid-kalaei Could you please make sure, that all Server-Dependencies are correctly and fully described here: https://github.com/kartevonmorgen/kartevonmorgen.ts#test-platform (Including OS, Node.js ...)
> 
> @flosse Can you please describe all Server-Requirements here https://github.com/kartevonmorgen/openfairdb/#build
> 
> Alex will provide Debian on his server and will give you test access. Do you think this could work?

the described requirements are the configuration for the current server and is performing very well for the testing purposes. for the production my estimation is it's also enough for production. at some point we may need to upgrade the computation power but in my eyes we are far from that time

**Comment by navid-kalaei** [comment]
> @navid-kalaei is it possible to start the new frontend with node.js instead of pm2 or can we create and use a Docker-Image?

`pm2` is just a process manager that watches for lifetime of the process, for example when the process stops unexpectedly or runs out of memory `pm2` restarts the process. so the answer is yes. we can run the project with `node`, however we should somehow be able to monitor the lifetime and the runtime and collect its logs.

Docker is very recommended however I have some questions:
- the build script halts at its last stage with the message of `server-side renders at runtime (uses getInitialProps or getServerSideProps)`. my guess is it's related to `SWC` as it's happening since I've updated Next version from 10 to 12. how we can terminate the build script when it reaches the end of the procedure?
- who will be responsible to manage the docker file and containers?
- docker is great for deployment, however it is complicated for junior developers, shall we write a tutorial for that?

**Comment by alex0107** [comment]
> * the build script halts at its last stage with the message of `server-side renders at runtime (uses getInitialProps or getServerSideProps)`. my guess is it's related to `SWC` as it's happening since I've updated Next version from 10 to 12. how we can terminate the build script when it reaches the end of the procedure?

I've tested with the linux command timeout but that doesn't work as expected...and I think that isn't a really good bugfix ;)

> * who will be responsible to manage the docker file and containers?

Maybe I? If there will be a new release, I can quickly upgrade the container or maybe we can use continuous deployment?

**Comment by navid-kalaei** [comment]
I found the cause of problem:

TL;DR:  [next-translate](https://github.com/vinissimus/next-translate) detects page props wrong. vinissimus/next-translate#486

Experience documentation:

we use [next-translate](https://github.com/vinissimus/next-translate) for the localization. 

1. the halting problem is related to the yarn's caching policy? to test that I have cleared the dependancies and cleared the cache but didn't work. here are the commands I run:
```
rm -rf node_modules
yarn cache clean
yarn install
yarn build:dev
```
2. on the build time, some warning pops out. is the halting related to the conflicts between dependencies or legacy modules? my answer is no because of two reasons. I always try to have dependencies updated and the message logs are warning not error, warnings are very usual and won't break projects. on the other hand, before migrating to Next 12 we did not break dependencies.

my hypothesis was related to upgrading Next from 10 to 12. which happened to be wrong

**Comment by alex0107** [comment]
After testing I found out the source-code is only compatible with Debian buster and Node 16 (Node 17 and / or bullseye not working)

**Comment by navid-kalaei** [comment]
@alex0107 there's a possibility of duplication at issue #70 . where are you going to discuss further? should we consider the build process closed and continue the server management here?

**Comment by alex0107** [comment]
@navid-kalaei sorry, I opened the issue before you have answered in this issue.

**Comment by wellemut** [comment]
Three Problems
1. Process of compiling did not complete, because of the package "next translate"
2. The Linux-Distribution of Alex was not working. Now, with Debian it is working
3. The Node Version 17 is not working, only 16 LTS (LongTermService Release)

**Comment by wellemut** [comment]
@alex0107 did you already manage to deploy the frontend somewhere?

**Comment by navid-kalaei** [comment]
> Three Problems
> 
> 1. Process of compiling did not complete, because of the package "next translate"
> 2. The Linux-Distribution of Alex was not working. Now, with Debian it is working
> 3. The Node Version 17 is not working, only 16 LTS (LongTermService Release)

just to mention the problems are already solved

**Comment by wellemut** [comment]
@alex0107 please send me your server IP, so I can set DNS entry for a test-URL to your server.
Most important for now is our new frontend. We can manage the backend later.

**Comment by wellemut** [comment]
- [ ] @alex0107 sends a pull request from https://github.com/kartevonmorgen/docker-images/blob/master/Frontend/Dockerfile-prod to this repo
- [ ] @navid-kalaei writes the github action and tries to run it via ssh (with the server scecrets from alex)

**Comment by navid-kalaei** [comment]
- the docker pull request is merged on commit 54da13abeaeed3a3120f24d266de7289122c284c on pr #83 
- the GitHub Actions with pm2 were integrated on Jan 14, 2022 at commit 162f51670702ee28d44f29cbda107c120a32b3e8 . the wiki is available at [How to deploy: self hosted GitHub Action](https://github.com/kartevonmorgen/kartevonmorgen.ts/wiki/How-to-deploy:-self-hosted-GitHub-Action). the credentials should be configured by the server owner. please don't hesitate to contact me if I could be any of assistance.

**Comment by wellemut** [comment]
The new server now (new.kartevonmorgen.org) runs on Hetzner and is Managed by @alex0107
